logstash-core/lib/logstash/pipeline.rb

```diff
@@ -473,10 +532,25 @@ def shutdown(&before_stop)
 
     before_stop.call if block_given?
 
+    stop_inputs
+
+    # We make this call blocking, so we know for sure when the method return the shtudown is
+    # stopped
+    wait_for_workers
+    clear_pipeline_metrics
+  end # def shutdown
+
+  def wait_for_workers
+    @logger.debug("Closing inputs")
```

Comment:
The biggest problem is we are not carrying any pipeline specific information, I have added that see my previous comment.